python/ray/serve/backend_state_manager.py

```diff
@@ -24,12 +24,51 @@
 SLOW_STARTUP_WARNING_PERIOD_S = 30
 
 
+class DeploymentState(Enum):
+    """
+    One abstraction level above replica to manage current state at
+    deployment level.
+
+    A deployment needs to keep track of two sets of states all time:
+      - (1) Target # of replicas running version X
+      - (2) Current # of replicas running version X
+
+    End user submits actions to change (1) through serve client,
+    such as:
+        - Deploying new deployment
+        - Add / remove replicas
+        - Update model version
+
+    Difference between (1) and (2) Determined current state of the
+    deployment that guides subsequent replica level actions.
+    """
+    # 0 replicas of the target state running
+    PENDING = 0
+    # Constructor failed N times for ALL replicas with 0 running,
+    # indicating last deployment action failed, will delete/rollback.
+    FAILED = 1
+    # > 1 but < target replica of the target state running, deployment
+    # is able to at least serve traffic
+    # All states below requires deployment to be at least in RUNNING
+    RUNNING = 2
+    # Recived request to deploy models at different version
+    UPGRADING = 3
+    # Adding more replicas to deployment
+    SCALING_UP = 4
+    # Removing replicas from deployment
+    SCALING_DOWN = 5
+    RUNNING_HEALTHY = 6
```

Comment:
Hmm what's the difference between this and `RUNNING`?

Reply:
Running means the deployment is available with at least 1 replica running with desired model version (technically we can also return deploy() so it doesn't block user code, and let controller to scale to target # eventually); while RUNNING_HEALTHY suggests it reached final desirable state such that no further action needed.

1) They can have better names and it's mostly draft for now 
2) They don't necessarily need to exposed to user, but rather an enum to keep track of total possible internal states of a running deployment, which we already sort of do in the logic of update()